File: apps/visualizer/backend/api/schemas/catalog.py
"""Catalog browse, similarity, and image-detail API models."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

class IdentityPerPerspectiveScore(BaseModel):
    model_config = ConfigDict(extra='forbid')

    perspective_slug: str
    display_name: str
    score: int
    prompt_version: str
    model_used: str
    scored_at: str
    rationale_preview: str


class CatalogImage(BaseModel):
    """Catalog list / search row shape (``query_catalog_images`` + API transforms)."""

    model_config = ConfigDict(extra='forbid')

    key: str
    id: int | None = None
    filename: str | None = None
    filepath: str | None = None
    date_taken: str | None = None
    rating: int | None = None
    pick: bool | None = None
    color_label: str | None = None
    keywords: list[str] = Field(default_factory=list)
    title: str | None = None
    caption: str | None = None
    description: str | None = None
    copyright: str | None = None
    width: int | None = None
    height: int | None = None
    instagram_posted: bool | None = None
    instagram_url: str | None = None
    image_hash: str | None = None
    image_type: Literal['catalog', 'instagram'] | None = None
    ai_analyzed: bool | None = None
    description_summary: str | None = None
    description_best_perspective: str | None = None
    description_perspectives: dict[str, Any] | None = None
    catalog_perspective_score: int | None = None
    catalog_score_perspective: str | None = None
    stack_id: int | None = None
    stack_member_count: int | None = None
    is_stack_representative: bool | None = None
    analyzed_at: str | None = None
    aperture: str | None = None
    camera_make: str | None = None
    camera_model: str | None = None
    catalog_path: str | None = None
    exif: str | None = None
    file_size: int | None = None
    focal_length: str | None = None
    gps_latitude: float | None = None
    gps_longitude: float | None = None
    instagram_index: int | None = None
    instagram_post_date: str | None = None
    iso: str | None = None
    lens: str | None = None
    phash: str | None = None
    shutter_speed: str | None = None
    similarity: float | None = None
    why_matched: str | None = None
    thumbnail_url: str | None = None
    score: float | None = None


class CatalogListResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    total: int
    images: list[CatalogImage]


class CatalogMonthsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    months: list[str]


class ClipSimilarMeta(BaseModel):
    model_config = ConfigDict(extra='forbid')

    clip_model_id: str
    clip_embed_dim: int
    knn_fetched: int
    knn_k_used: int | None = None


class CatalogSimilarResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    images: list[CatalogImage]
    total: int
    meta: ClipSimilarMeta


class CatalogSimilarityGroup(BaseModel):
    model_config = ConfigDict(extra='forbid')

    group_id: int
    seed: CatalogImage
    candidates: list[CatalogImage]
    candidate_count: int
    best_similarity: float
    job_id: str | None = None
    created_at: str | None = None


class CatalogSimilarityGroupsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    items: list[CatalogSimilarityGroup]
    total: int


class ImageView(BaseModel):
    """``GET /api/images/<image_type>/<image_key>`` consolidated detail shape."""

    model_config = ConfigDict(extra='forbid')

    image_type: Literal['catalog', 'instagram']
    key: str
    id: int | None = None
    filename: str | None = None
    filepath: str | None = None
    local_path: str | None = None
    date_taken: str | None = None
    created_at: str | None = None
    rating: int | None = None
    pick: bool | None = None
    color_label: str | None = None
    keywords: list[str] | None = None
    title: str | None = None
    caption: str | None = None
    copyright: str | None = None
    width: int | None = None
    height: int | None = None
    instagram_posted: bool | None = None
    instagram_url: str | None = None
    post_url: str | None = None
    image_hash: str | None = None
    stack_id: int | None = None
    stack_member_count: int | None = None
    is_stack_representative: bool | None = None
    instagram_folder: str | None = None
    date_folder: str | None = None
    source_folder: str | None = None
    matched_catalog_key: str | None = None
    processed: bool | None = None
    ai_analyzed: bool | None = None
    description_summary: str | None = None
    description_best_perspective: str | None = None
    catalog_perspective_score: int | None = None
    catalog_score_perspective: str | None = None
    available_score_perspectives: list[str] | None = None
    identity_aggregate_score: float | None = None
    identity_perspectives_covered: int | None = None
    identity_eligible: bool | None = None
    identity_per_perspective: list[IdentityPerPerspectiveScore] = Field(default_factory=list)
    analyzed_at: str | None = None
    aperture: str | None = None
    camera_make: str | None = None
    camera_model: str | None = None
    catalog_path: str | None = None
    description: str | None = None
    exif: str | None = None
    exif_data: Any | None = None
    file_size: int | None = None
    focal_length: str | None = None
    gps_latitude: float | None = None
    gps_longitude: float | None = None
    instagram_index: int | None = None
    instagram_post_date: str | None = None
    iso: str | None = None
    lens: str | None = None
    phash: str | None = None
    shutter_speed: str | None = None
    added_at: str | None = None
    file_path: str | None = None
    last_attempted_at: str | None = None
    media_key: str | None = None
    processed_at: str | None = None
    vision_result: str | None = None
    vision_score: float | None = None


ImageDetailResponse = ImageView


def validate_catalog_image(row: dict) -> dict:
    return CatalogImage.model_validate(row).model_dump(mode='json')


def validate_image_view(row: dict) -> dict:
    return ImageView.model_validate(row).model_dump(mode='json')

File: apps/visualizer/backend-ts/src/api/schemas/catalog.ts
/**
 * Catalog browse, similarity, and image-detail API models.
 * Port of `api/schemas/catalog.py`.
 *
 * Every model here is `.strict()`, matching pydantic's `extra='forbid'`. That is
 * load-bearing rather than stylistic: the catalog row is built by spreading a
 * `SELECT i.*` result, so a column added to `images` would otherwise leak into the
 * API and into `api.gen.ts` without anyone deciding to expose it.
 */
import { z } from '@hono/zod-openapi';

export const IdentityPerPerspectiveScore = z
  .object({
    perspective_slug: z.string(),
    display_name: z.string(),
    score: z.int(),
    percentile: z.number().nullish(),
    prompt_version: z.string(),
    model_used: z.string(),
    scored_at: z.string(),
    rationale_preview: z.string(),
  })
  .strict()
  .openapi('IdentityPerPerspectiveScore');

/**
 * Catalog list / search row shape (`queryCatalogImages` plus the API transforms).
 *
 * The field order follows the Python model so the generated schema diffs cleanly.
 */
export const CatalogImage = z
  .object({
    key: z.string(),
    id: z.int().nullish(),
    filename: z.string().nullish(),
    filepath: z.string().nullish(),
    date_taken: z.string().nullish(),
    rating: z.int().nullish(),
    pick: z.boolean().nullish(),
    color_label: z.string().nullish(),
    keywords: z.array(z.string()).default([]),
    title: z.string().nullish(),
    caption: z.string().nullish(),
    description: z.string().nullish(),
    copyright: z.string().nullish(),
    width: z.int().nullish(),
    height: z.int().nullish(),
    instagram_posted: z.boolean().nullish(),
    image_hash: z.string().nullish(),
    image_type: z.literal('catalog').nullish(),
    ai_analyzed: z.boolean().nullish(),
    description_summary: z.string().nullish(),
    description_best_perspective: z.string().nullish(),
    catalog_perspective_score: z.int().nullish(),
    catalog_score_perspective: z.string().nullish(),
    stack_id: z.int().nullish(),
    stack_member_count: z.int().nullish(),
    is_stack_representative: z.boolean().nullish(),
    analyzed_at: z.string().nullish(),
    aperture: z.string().nullish(),
    camera_make: z.string().nullish(),
    camera_model: z.string().nullish(),
    catalog_path: z.string().nullish(),
    exif: z.string().nullish(),
    file_size: z.int().nullish(),
    focal_length: z.string().nullish(),
    gps_latitude: z.number().nullish(),
    gps_longitude: z.number().nullish(),
    iso: z.string().nullish(),
    lens: z.string().nullish(),
    phash: z.string().nullish(),
    shutter_speed: z.string().nullish(),
    similarity: z.number().nullish(),
    why_matched: z.string().nullish(),
    thumbnail_url: z.string().nullish(),
    score: z.number().nullish(),
  })
  .strict()
  .openapi('CatalogImage');

export const InstagramPostedRequest = z
  .object({ posted: z.boolean() })
  .strict()
  .openapi('InstagramPostedRequest');

export const InstagramPostedResponse = z
  .object({ key: z.string(), instagram_posted: z.boolean() })
  .strict()
  .openapi('InstagramPostedResponse');

export const CatalogListResponse = z
  .object({ total: z.int(), images: z.array(CatalogImage) })
  .strict()
  .openapi('CatalogListResponse');

export const CatalogMonthsResponse = z
  .object({ months: z.array(z.string()) })
  .strict()
  .openapi('CatalogMonthsResponse');

export const ClipSimilarMeta = z
  .object({
    clip_model_id: z.string(),
    clip_embed_dim: z.int(),
    knn_fetched: z.int(),
    knn_k_used: z.int().nullish(),
  })
  .strict()
  .openapi('ClipSimilarMeta');

export const CatalogSimilarResponse = z
  .object({
    images: z.array(CatalogImage),
    total: z.int(),
    meta: ClipSimilarMeta,
  })
  .strict()
  .openapi('CatalogSimilarResponse');

export const CatalogSimilarityGroup = z
  .object({
    group_id: z.int(),
    seed: CatalogImage,
    candidates: z.array(CatalogImage),
    candidate_count: z.int(),
    best_similarity: z.number(),
    job_id: z.string().nullish(),
    created_at: z.string().nullish(),
  })
  .strict()
  .openapi('CatalogSimilarityGroup');

export const CatalogSimilarityGroupsResponse = z
  .object({ items: z.array(CatalogSimilarityGroup), total: z.int() })
  .strict()
  .openapi('CatalogSimilarityGroupsResponse');

/**
 * `GET /api/images/catalog/{image_key}` consolidated detail shape.
 *
 * Wider than `CatalogImage` and not a superset of it: the detail payload adds the
 * identity block and the raw `images` columns the grid does not need, and it drops
 * the similarity fields the grid adds. Several fields (`instagram_folder`,
 * `media_key`, `vision_result`) are vestiges of the retired Instagram scope, kept
 * because the frontend's generated types still reference the shape.
 */
export const ImageView = z
  .object({
    image_type: z.literal('catalog'),
    key: z.string(),
    id: z.int().nullish(),
    filename: z.string().nullish(),
    filepath: z.string().nullish(),
    local_path: z.string().nullish(),
    date_taken: z.string().nullish(),
    created_at: z.string().nullish(),
    rating: z.int().nullish(),
    pick: z.boolean().nullish(),
    color_label: z.string().nullish(),
    keywords: z.array(z.string()).nullish(),
    title: z.string().nullish(),
    caption: z.string().nullish(),
    copyright: z.string().nullish(),
    width: z.int().nullish(),
    height: z.int().nullish(),
    instagram_posted: z.boolean().nullish(),
    post_url: z.string().nullish(),
    image_hash: z.string().nullish(),
    stack_id: z.int().nullish(),
    stack_member_count: z.int().nullish(),
    is_stack_representative: z.boolean().nullish(),
    instagram_folder: z.string().nullish(),
    date_folder: z.string().nullish(),
    source_folder: z.string().nullish(),
    matched_catalog_key: z.string().nullish(),
    processed: z.boolean().nullish(),
    ai_analyzed: z.boolean().nullish(),
    description_summary: z.string().nullish(),
    description_best_perspective: z.string().nullish(),
    catalog_perspective_score: z.int().nullish(),
    catalog_score_perspective: z.string().nullish(),
    available_score_perspectives: z.array(z.string()).nullish(),
    identity_aggregate_score: z.number().nullish(),
    identity_peak_percentile: z.number().nullish(),
    identity_perspectives_covered: z.int().nullish(),
    identity_eligible: z.boolean().nullish(),
    identity_per_perspective: z.array(IdentityPerPerspectiveScore).default([]),
    analyzed_at: z.string().nullish(),
    aperture: z.string().nullish(),
    camera_make: z.string().nullish(),
    camera_model: z.string().nullish(),
    catalog_path: z.string().nullish(),
    description: z.string().nullish(),
    exif: z.string().nullish(),
    exif_data: z.unknown().nullish(),
    file_size: z.int().nullish(),
    focal_length: z.string().nullish(),
    gps_latitude: z.number().nullish(),
    gps_longitude: z.number().nullish(),
    iso: z.string().nullish(),
    lens: z.string().nullish(),
    phash: z.string().nullish(),
    shutter_speed: z.string().nullish(),
    added_at: z.string().nullish(),
    file_path: z.string().nullish(),
    last_attempted_at: z.string().nullish(),
    media_key: z.string().nullish(),
    processed_at: z.string().nullish(),
    vision_result: z.string().nullish(),
    vision_score: z.number().nullish(),
  })
  .strict()
  .openapi('ImageView');

export type CatalogImage = z.infer<typeof CatalogImage>;
export type ImageView = z.infer<typeof ImageView>;
